### PR TITLE
Add shuffle capability in local mode

### LINKED ISSUE #243: Add shuffle support for reshape with odd shapes in threaded mode
**Is your feature request related to a problem? Please describe.**
According to #134 , we need to add shuffle support for ``reshape`` with odd shapes. This provides tensor-level paradigm for shuffle operations.

**Describe the solution you'd like**
A private argument is added to ``reshape`` function as an option to enable reshape by shuffle. As distributed mode not adopted, the argument shall not break existing code. When tests are completed, this argument can be removed.

### PR DESCRIPTION
**Pull request by wjsi**
## What do these changes do?

Add shuffle functionality for ```reshape``` with irregular shape. Note that this behavior is enabled only with bare ```reshape``` function with ```_reshape_with_shuffle``` arg, thus previous behavior is not broken.

## Related issue number

Fixes #243

### CODE REVIEW

File: mars/tensor/expressions/tests/test_reshape.py [hunk 1]
```diff
@@ -48,3 +49,18 @@ def testReshape(self):
         self.assertEqual(calc_shape(b), b.shape)
         self.assertEqual(calc_shape(b.chunks[0]), b.chunks[0].shape)
         self.assertEqual(tuple(sum(s) for s in a.nsplits), (10, 30, 20))
+
+    def testShuffleReshape(self):
+        from mars.tensor import reshape
+        a = ones((31, 27), chunk_size=10)
+        b = reshape(a, (27, 31), _reshape_with_shuffle=True)
```
Comment (chineking) [inline mars/tensor/expressions/tests/test_reshape.py]: `_reshape_with_shuffle` looks pretty weird, if it's only for tests purpose, just modify the op property in the tests looks fine to me.

File: mars/tensor/expressions/reshape/reshape.py [hunk 2]
```diff
@@ -154,6 +157,40 @@ def _reduce_nsplit_by_expand(splits, expand):
...
+        tensor = op.outputs[0]
+        new_shape = op.newshape
+        shuffle_inputs, shuffle_outputs = [], []
+        axis_offsets = [[0] + np.cumsum(ns)[:-1].tolist() for ns in in_tensor.nsplits]
+
+        chunk_size_tuples = tuple((max(tp), len(tp)) for tp in in_tensor.nsplits)
+        max_chunk_size = max(tp[0] for tp in chunk_size_tuples)
+        chunk_size = tuple(tp[0] if tp[1] > 1 else max_chunk_size
+                           for tp in chunk_size_tuples)
+        out_nsplits = decide_chunk_sizes(new_shape, chunk_size, tensor.dtype.itemsize)
```
Comment (chineking) [inline mars/tensor/expressions/reshape/reshape.py]: If the reshaped tensor has different `ndim` with the input tensor, some errors may occur here.

**Comment by chineking** [comment]
I think we can merge this PR first after modification, and then try to optimize the execution of `TensorReshapeMap` and `TensorReshapeReduce` later. After all, it's not easy to optimize the performance.

File: mars/tensor/expressions/core.py [hunk 3]
```diff
@@ -54,3 +59,13 @@ def new_tensor(self, inputs, shape, dtype=None, **kw):
 
     def calc_shape(self, *inputs_shape):
         raise NotImplementedError
+
+
+class TensorShuffleProxy(ShuffleProxy, TensorOperandMixin):
+    def __init__(self, **kwargs):
+        if '_proxy_id' not in kwargs:
+            kwargs['_proxy_id'] = random.randint(0, 0x7fffffff)
```
[the remark this answers is not in the document]
Comment (wjsi) [reply mars/tensor/expressions/core.py]: Removed and use ``op.params`` instead.

File: mars/tensor/execution/reshape.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+def _reshape_map(ctx, chunk):
+    old_shape = chunk.op.oldshape
+    new_shape = chunk.op.newshape
+    new_chunk_size = chunk.op.new_chunk_size
+    axis_offset = chunk.op.axis_offsets
+
+    logger.debug('Reshape mapper: Start mapping step for %s', chunk.key)
+
+    data = ctx[chunk.inputs[0].key]
```
[the remark this answers is not in the document]
Comment (chineking) [reply mars/tensor/execution/reshape.py]: We shall do this in the coming performance tuning, resolve for now.

File: mars/tensor/execution/reshape.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    indices = list(np.nonzero(data))
+    nz_data = data[indices]
```
[the remark this answers is not in the document]
Comment (chineking) [reply mars/tensor/execution/reshape.py]: As `indices` is used further, this suggestion may not be critical. Resolve first.

File: mars/tensor/expressions/reshape/reshape.py [hunk 6]
```diff
@@ -178,11 +213,66 @@ def tile(cls, op):
             return new_op.new_tensors(op.inputs, tensor.shape,
                                       chunks=out_chunks, nsplits=reshape_nsplits)
         except ValueError:
+            # TODO: make this as sefault when shuffle is mature
```
Comment (NJUPole) [inline mars/tensor/expressions/reshape/reshape.py]: sefault -> default

**Pull request merged**